### DISCUSSION TITLE
add a custom shader to offscreenrender

### DISCUSSION THREAD

**Discussion by WdsPjm** [vsg-dev/VulkanSceneGraph#1024] [Q&A] [answered]
In vsgrendertotexture, I want to add a custom shader to offscreenrender, but after that it only works on the main_RenderGraph.

**Comment by robertosfield** [comment, marked as answer]
Back in August I implemented support fro selecting GraphicsPipline based in viewID so that each View could have it's own GraphicsPipeline, and then I added support for building seperate GraphicsPipeline per View by adding mask support into ShaderStage and GraphicsPipelineState which is used when creating the required vkPipeline objects for these.  The mask on the ShaderState & GraphicsPipeineState is matched to each View's mask.  

I wrote up this work on this thread: #916

In a custom ShaderSet you can add the ShaderStage and GraphicsPipelineState with the approach masks.

**Comment by WdsPjm** [comment]
It's very useful

**Comment by WdsPjm** [comment]
I'm testing at vsganaglyphicstereo demo.

void enableCustomShaderSets(vsg::Mask leftMask, vsg::Mask rightMask, vsg::ref_ptr<vsg::Options> options)
{
    auto left_colorBlendState = vsg::RasterizationState::create();
    left_colorBlendState->mask = leftMask;
    left_colorBlendState->polygonMode = VK_POLYGON_MODE_LINE;
    

    auto right_colorBlendState = vsg::RasterizationState::create();
    right_colorBlendState->mask = rightMask;
   

    right_colorBlendState->polygonMode = VK_POLYGON_MODE_FILL;;
    vsg::GraphicsPipelineStates pipelineStates{left_colorBlendState, right_colorBlendState};


    vsg::ref_ptr<vsg::ShaderSet> shaderSet = pbr_ShaderSet(options);
    options->shaderSets["pbr"] = shaderSet;

    auto& pbr = options->shaderSets["pbr"];
    pbr->defaultGraphicsPipelineStates = pipelineStates;
}


A custom shader is used here   The shaderstate in the custom shader also sets the corresponding mask

vsg::Mask InMask = 0x1;
vsg::ref_ptr<vsg::ShaderSet> pbr_ShaderSet(vsg::ref_ptr<const vsg::Options> options)
{
    vsg::info("Local pbr_ShaderSet(", options, ")");

    auto vertexShader = vsg::read_cast<vsg::ShaderStage>("shaders/pickOffscreen.vert", options);
    auto fragmentShader = vsg::read_cast<vsg::ShaderStage>("shaders/pickOffscreen.frag", options);
    vertexShader->mask = InMask;
    fragmentShader->mask = InMask;
    if (!vertexShader || !fragmentShader)
    {
        vsg::error("pbr_ShaderSet(...) could not find shaders.");
        return {};
    }


But both pipelines use custom shaders



![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/52599568/bbf65911-f40f-4dc6-beb2-c66330e496f5)



Am I using the wrong method

**Reply by robertosfield** [reply]
You explain a bit of what you are doing, but don't say what end result you are going for so really can't comment on whether you are doing what is required.

**Comment by WdsPjm** [comment]
I want both pipelines to have their own shaders

**Reply by robertosfield** [reply]
You'll need to assign all the shadows to the same ShaderSet but set the mask on each ShaderStage to specify which view mask it should be used on, so the original ShaderStage will need their own mask values set, as well as the new ShaderStage that you are adding.

**Comment by WdsPjm** [comment]
I'm sorry. I didn't make it clear what I was going to do. In the vsgrendertotexture example, offscreenrenderpass and mainrenderpass use the same default shader. I want to add a custom shader to offscreenrenderpass. mainrenderpass uses the default shader. I will assign ids to all the mesh bodies through the offscreenRenderpass channel, and then color the mesh bodies based on the ids.

**Reply by robertosfield** [reply]
As i have explain already, you will want to use a single custom ShaderSet for the main View and offscreen View and differentiate between the two using a vsg::Msak bit dedicated to the main and the off screen view, and then match this Mask bits to the ones uses when setting up the shaders for the main View and the offscreen VIew.

**Comment by WdsPjm** [comment]
auto left_colorBlendState = vsg::RasterizationState::create();
    left_colorBlendState->mask = leftMask;
    left_colorBlendState->polygonMode = VK_POLYGON_MODE_LINE;
 
    auto right_colorBlendState = vsg::RasterizationState::create();
    right_colorBlendState->mask = rightMask;
    right_colorBlendState->polygonMode = VK_POLYGON_MODE_FILL;;
    vsg::GraphicsPipelineStates pipelineStates{left_colorBlendState, right_colorBlendState};


    vsg::ref_ptr<vsg::ShaderSet> shaderSetLeft = pbr_ShaderSetLeft(options);
    auto& pbrLeft = options->shaderSets["pbr"] = shaderSetLeft;
    for (auto& t : pbrLeft->stages)
    {
        t->mask = leftMask;
    }
    pbrLeft->defaultGraphicsPipelineStates = pipelineStates;


    vsg::ref_ptr<vsg::ShaderSet> shaderSetRight = pbr_ShaderSetRight(options);
    auto& pbrRight = options->shaderSets["pbr"] = shaderSetRight;
    for (auto& t : pbrRight->stages)
    {
        t->mask = rightMask;
    }

  pbrRight->defaultGraphicsPipelineStates = pipelineStates;


I used two custom Shadersets, shadersetLeft and shadersetRight, was I wrong so that only one shaderset worked




![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/52599568/00403be6-0b05-4ab4-850f-faecfabbd0d2)

**Comment by robertosfield** [comment]
Without all the code in front of my I don't know exactly what you are doing.  However, as I have been saying all along your should use one ShaderSet as you are building on subgraph that will be rendering in both views.

The key to having different rendering in each view is to use the ShaderStage::mask and GraphicsPipelineState::mask parameters to differentiate what will be used for the graphics pipeline  used in each view.  

The vsganaglphhicstereo example illustrastates most of the key parts of what is required - one ShaderSet with multiple GraphicsPipelineState that use the mask to specialize it for each view.  The only addition beyond this you need to make is set the ShaderStage::mask so adding multiple shaders stages to the single ShaderSet will map to appropriate ShaderSet stage used for the desired View.

I can't keep saying this same thing over and over.  Please read what I've written, and the vsganalgphicstereo example to see the illustration of what I'm talking about.

**Comment by WdsPjm** [comment]
Steps as you said, I am in vsganalgphicstereo, add a pbr_ShaderSetLeft replacement for VSG: : createPhysicsBasedRenderingShaderSet (options); In pbr_ShaderSetLeft, leftMask is set for each of the two Shaderstages

vsg::info("Local pbr_ShaderSet(", options, ")");

auto vertexShader = vsg::read_cast<vsg::ShaderStage>("shaders/pickOffscreen.vert", options);
auto fragmentShader = vsg::read_cast<vsg::ShaderStage>("shaders/pickOffscreen.frag", options);

vsg::Mask leftMask = 0x1;
vertexShader->mask = leftMask;
fragmentShader->mask = leftMask;


The result is that leftview and rightview have various GraphicsPipelineStates, but they both use pbr_ShaderSetLeft. What I want is that only leftview uses pbr_ShaderSetLeft.

**Comment by robertosfield** [comment]
I don;t know how many ways I can say the same thing,

You can only use one ShaderSet for building a single subgraph, that one
ShaderSet will need the left and right varians of the ShaderSet and
GraphicsPiplineState,  Again follow the vsganaglyjpicsstereo example, it
doesn't have two ShaderSet, just one that is specialized for each View.

There is no left ShaderSet, right ShaderSet, there's is just one ShaderSet
that is specialized for each VIew based on the View' mask being matched to
the mask of the settings assigned to that ShaderSet.

**Comment by WdsPjm** [comment]
I am very sorry that I misunderstood your advice from the very beginning. Now it works properly

<img width="1520" alt="b4383ae1dc822c515e440d0c5c13f28" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/52599568/23a9c5c1-4d7e-466c-9c72-ba11f2c8627a">

**Reply by robertosfield** [reply]
Good to hear it's now working fine.
